### DISCUSSION TITLE
can i create a drop-in repl with variables in scope from preloaded files?

### DISCUSSION THREAD

**Discussion by adccb** [TypeStrong/ts-node#1280] [Q&A]
basically i have a new typescript project, and i'd love to be able to start up a repl and instantly have access to project types defined in a file. specifically, things defined like:

```typescript
export type Dict<T> = Record<string, T>;
export type StringDict = Dict<string>;
export type Collection<T> = T[] | Dict<T>;

```

i'd love to be able to say `ts-node -r src/types/index.ts` and have it create a repl where i can just use `Dict`. is this possible?

**Comment by adccb** [comment]
for clarity, i've tried setting up a `repl_setup.ts` that looks like this:

```ts
export * as types from './src/types/index'
```
but it doesn't look like these variables are populated in the repl at all—
```
> ts-node -r ./src/repl_setup.ts

> types.Maybe
[eval].ts:1:1 - error TS2304: Cannot find name 'types'.

1 types.Maybe
  ~~~~~
```

is there some other way to do this?

**Reply by cspotcode** [reply]
The REPL is essentially its own module, just like any other file.  If file A declares exports, then file B needs to import them to be able to use them.  Otherwise, they are not in scope.

It sounds like you want a REPL where a few import statements have automatically been run before the user is prompted for input.

You'll have to experiment, but this might work.  I have not tested it:

```
ts-node --interactive -e "import {Things} from './other-modules';"
```

If that works (a big if) then the expression will be executed in the REPL automatically, and then the user will be given interactive control to type additional lines of code.

Our REPL is also exposed via our API, so you can experiment with writing your own wrapper script that uses our API to instantiate and start a new REPL.  The usage example on the `createRepl` function should be enough to get you started: https://typestrong.org/ts-node/api/index.html#createrepl

**Reply by adccb** [reply]
just getting back to this now, thanks!

that `--interactive` didn't work, so i'm digging into scripting a repl myself—i can create the repl, and i can dynamic-`import()` to get actual concrete values, but i can't get the types over. i've tried both
```ts
repl.start(`import type { Thing } from '${typeRoot}'`)
// and
replService.evalCode(`import type { Thing } from '${typeRoot}'`)
repl.start()
```
and neither seems to get me there—they both fail with:
```
/$(pwd)/[eval].ts:2
Object.defineProperty(exports, "__esModule", { value: true });
                      ^

ReferenceError: exports is not defined
```

and i _think_ that means that the repl context is not a module. is there a way to import types from this context?

**Reply by adccb** [reply]
forgot to include my `tsconfig.json`
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "noUnusedLocals": false,
    "sourceMap": true
  }
}
```

**Reply by cspotcode** [reply]
I tested the `--interactive` idea, and it looks like you've uncovered a bug.  I create #1282 to track.  It looks like you are also seeing the bug when calling `evalCode` before `start`.  I will track all progress on a bugfix in #1282.

In the meantime, you *might* have luck postponing your `evalCode` call until *after* calling `start`.

**Reply by cspotcode** [reply]
@mxtetrachord I overhauled how we configure the environment in our REPL functionality.  Those changes were merged in #1333.  I logged detailed information about node's various `--eval`, stdin, and REPL environments, turned that information into test cases, and updated `ts-node` to match.

This will be released in the next version, hopefully soon, and in the meantime you can install from git.  I believe that the following should work for you:

```
repl.start();
repl.evalCode(`import * as Types from "${Path.resolve(__dirname, 'types')}";\n`);
```

Complete example is here: https://github.com/TypeStrong/ts-node-repros/tree/1280
Logs here: https://github.com/TypeStrong/ts-node-repros/runs/2777457756?check_suite_focus=true

Calling `start()` first is a limitation of node's REPL API, where the global environment is not configured until the REPL is started.  There may be API polish we can do from here; let me know if you have any feedback.  The changes in #1333 were by far the hardest part; any API improvements from here should be straightforward PRs for even new contributors.
